### PR TITLE
Enabling alternative controller rest API classes in preprocess

### PR DESCRIPTION
**Pull request by jenniferdai**
Adding ability to override ControllerRestAPI in custom implementations of SegmentPreprocessingJob

### CODE REVIEW

File: pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentCreationJob.java [hunk 1]
```diff
@@ -194,14 +193,6 @@ public void run()
...
 
-  @Nullable
-  protected TableConfig getTableConfig()
-      throws IOException {
-    try (ControllerRestApi controllerRestApi = getControllerRestApi()) {
-      return controllerRestApi != null ? controllerRestApi.getTableConfig() : null;
-    }
-  }
-
   protected Schema getSchema()
```
Comment (Jackie-Jiang) [inline pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentCreationJob.java]: Move this to Base class as well

File: pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java [hunk 2]
```diff
@@ -375,13 +377,37 @@ private void addHashCodeField(Set<Schema.Field> fieldSet) {
     fieldSet.add(hashCodeField);
   }
 
+  protected org.apache.pinot.common.data.Schema getSchema()
```
Comment (Jackie-Jiang) [inline pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java]: Use the one in Base class, remove the duplicate code in this class

File: pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentCreationJob.java (hunk 1 above)
Comment (jenniferdai) [reply pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentCreationJob.java]: The way we support schemas means we potentially need to initialize the filesystem and open the schema file. Without running the job (as this is an abstract class), we cannot do that.

File: pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java (hunk 2 above)
Comment (jenniferdai) [reply pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java]: same as above

File: pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/BaseSegmentJob.java [hunk 3]
```diff
@@ -37,13 +41,57 @@
...
+    }
+  }
+
+  /**
+   * This method is currently implemented in SegmentCreationJob and SegmentPreprocessingJob due to a dependency on
+   * the hadoop filesystem, which we can only get once the job begins to run.
+   * We return null here to make it clear that for now, all implementations of this method have to support
+   * reading from a schema file. In the future, we hope to deprecate reading the schema from the schema file in favor
+   * of mandating that a schema is pushed to the controller.
+   * @return
```
Comment (Jackie-Jiang) [inline pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/BaseSegmentJob.java]: Remove this line

File: pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentCreationJob.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentCreationJob.java]: Actually we can move _filesystem and _schemaFile into the base class as well. Your call

File: pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java [hunk 4]
```diff
@@ -375,13 +377,37 @@ private void addHashCodeField(Set<Schema.Field> fieldSet) {
...
+  }
+
   private void setTableConfigAndSchema() throws IOException {
     // If push locations, table config, and schema are not configured, this does not necessarily mean that segments
     // cannot be created. We should allow the user to go to the next step rather than failing the job.
-    Preconditions.checkState(!_pushLocations.isEmpty(), "Push locations cannot be empty.");
-    try(ControllerRestApi controllerRestApi = new DefaultControllerRestApi(_pushLocations, _rawTableName)) {
-      _tableConfig = controllerRestApi.getTableConfig();
-      _pinotTableSchema = controllerRestApi.getSchema();
+    try(ControllerRestApi controllerRestApi = getControllerRestApi()) {
```
Comment (Jackie-Jiang) [inline pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java]: Here we don't allow using `_schemaFile`?

Comment (jenniferdai) [reply pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java]: Changed

File: pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/BaseSegmentJob.java [hunk 5]
```diff
@@ -37,13 +41,56 @@
...
+    String pushPortString = _properties.getProperty(JobConfigConstants.PUSH_TO_PORT);
+    if (pushHostsString != null && pushPortString != null) {
+      _pushLocations =
+          PushLocation.getPushLocations(StringUtils.split(pushHostsString, ','), Integer.parseInt(pushPortString));
+    } else {
+      _pushLocations = null;
+    }
+
+    _rawTableName = Preconditions.checkNotNull(_properties.getProperty(JobConfigConstants.SEGMENT_TABLE_NAME));
+
```
Comment (Jackie-Jiang) [inline pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/BaseSegmentJob.java]: (nit) remove empty line

File: pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java [hunk 6]
```diff
@@ -87,10 +88,11 @@
   private final Path _inputSegmentDir;
   private final Path _preprocessedOutputDir;
   protected final String _rawTableName;
-  protected final List<PushLocation> _pushLocations;
+  protected List<PushLocation> _pushLocations;
```
Comment (Jackie-Jiang) [inline pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java]: Remove the member variables that are already in the base segment job (_pushLocations and _rawTableName). Same for SegmentCreationJob.

File: pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java [hunk 7]
```diff
@@ -115,9 +119,7 @@ public SegmentPreprocessingJob(final Properties properties) {
       _pushLocations =
           PushLocation.getPushLocations(StringUtils.split(pushHostsString, ','), Integer.parseInt(pushPortString));
     } else {
-      throw new RuntimeException(String
-          .format("Push location is mis-configured! %s: %s, %s: %s", JobConfigConstants.PUSH_TO_HOSTS, pushHostsString,
-              JobConfigConstants.PUSH_TO_PORT, pushPortString));
+      _pushLocations = null;
```
Comment (Jackie-Jiang) [inline pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java]: Remove the code of setting _pushLocations and _rawTableName which are already in the BaseSegmentJob. Same for SegmentCreationJob.

**Pull request merged**